### ISSUE TITLE
Chinese font problem

### ISSUE THREAD

**Issue by jingjinghack** [ocornut/imgui#1434]
```cpp
ImGuiIO& io = ImGui::GetIO();
io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\simhei.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());
```
This takes up a lot of memory(300MB)
it just a font...
![tim 20171114085133](https://user-images.githubusercontent.com/16219523/32770890-14a42e8c-c95c-11e7-96d3-2225db5a97dd.png)

**Comment by ratchetfreak** [comment]
The font file is 9 MB. Which is compressed. To make efficient use out of the font it needs to be decompressed. 

IMO it is not unsurprising that a font file converted to usable data structure explodes 30-fold especially when the render method is bitmap based.

**Comment by ocornut** [comment]
With the current system of baking full ranges into a texture atlas, if I load a test font at 18 px size with full Chinese range + default oversampling setting, I get a 4096x4096 texture. You can see it in the Demo Window under Windows Options > Font.

A 4096x4096 texture in RGBA with will be 64 MB of RAM + Alpha texture currently stays stored that's 16 MB. It's too much but it's not 300 MB.

![image](https://user-images.githubusercontent.com/8225057/32773603-fa0c6d78-c929-11e7-81d5-9383a5435f91.png)

The three solutions are:
- You can reduce the oversampling.
```
    ImFontConfig cfg;
    cfg.OversampleH = cfg.OversampleV = 1;
    ImFont* font = io.Fonts->AddFontFromFileTTF("c:\\Omar\\WB\\work\\game\\data\\font\\NotoSansCJKsc-Medium.otf", 18.0f, &cfg, io.Fonts->GetGlyphRangesChinese());
```
Texture becomes 4096x1024 so 4 MB alpha and 16 MB RGBA.

- You can reduce the glyphs range by computing your own ranges for the characters you actually need to display (this work if you text is known). This also will be much faster to build.
The `ImFontAtlas::GlyphRangesBuilder` class is a helper to build ranges given input tetxt.

- You can wait until the atlas system gets support for dynamically adding characters as they are needed (probably not before several months).

**Comment by jingjinghack** [comment]
@ocornut i use 1nd solution, thanks for answers

**Comment by jingjinghack** [comment]
@ratchetfreak how to do i uncompress it?

**Comment by ratchetfreak** [comment]
@jingjinghack that's what AddFontFromFileTTF does, it decompresses the file into a usable format and then adds the required data structures to the text renderer.

**Comment by ocornut** [comment]
FYI, work on `dynamic_fonts` branch aimed to be merged in 1.92 (#8465) should solve this, as glyphs are rasterized and packed on demand when they are used.
